packages/hydrooj/src/model/user.ts

```diff
@@ -161,14 +165,16 @@ export class User {
 
     serialize(h) {
         if (!this._isPrivate) {
-            const fields = ['_id', 'uname', 'mail', 'perm', 'role', 'priv', 'regat', 'loginat', 'tfa', 'authn'];
-            if (h?.user?.hasPerm(PERM.PERM_VIEW_DISPLAYNAME)) {
-                fields.push('displayName', 'school', 'studentId');
-            }
+            const fields = [...new Set(User._allowedFields)];
+            if (h?.user?.hasPerm(PERM.PERM_VIEW_DISPLAYNAME)) fields.push('displayName', 'school', 'studentId');
             return pick(this, fields);
         }
         return JSON.stringify(this, serializer(true, h));
     }
+
+    static addAllowedFields(...fields: string[]) {
+        User._allowedFields.push(...fields);
```

Comment:
side effects should be implemented as disposables (check how other services do garbage collection)

also likely needs to dedupe those fields